File: 1458-sort-integers-by-the-number-of-1-bits/1458-sort-integers-by-the-number-of-1-bits.java
class Solution {
    public class pair{
        int num;
        int no1;
        pair(){}
        pair(int num, int no1){
            this.num = num;
            this.no1 = no1;
        }
    }
    public int[] sortByBits(int[] arr) {
        PriorityQueue<pair> pq = new PriorityQueue<>((a,b)->{
            if(a.no1 == b.no1){
                return a.num - b.num;
            }
            return a.no1 - b.no1;
        });

        for(int val : arr){
            int cnt = 0;
            for(int i=0; i<32; i++){
                int bit = ((1<<i) & val) == 0 ? 0 : 1;
                if(bit == 1){
                    cnt++;
                }
            }
            pq.add(new pair(val, cnt));
        }

        int[] ans = new int[arr.length];
        int ind = 0;
        while(pq.size() != 0){
            pair rem = pq.remove();
            ans[ind++] = rem.num;
        }

        return ans;
    }
}

File: 1458-sort-integers-by-the-number-of-1-bits/1458-sort-integers-by-the-number-of-1-bits.cpp
class Solution {
public:
    struct Pair {
        int num;
        int no1;

        Pair(int num, int no1) : num(num), no1(no1) {}
        
        bool operator>(const Pair& other) const {
            if (no1 == other.no1) {
                return num > other.num;
            }
            return no1 > other.no1;
        }
    };

    std::vector<int> sortByBits(std::vector<int>& arr) {
        std::priority_queue<Pair, std::vector<Pair>, std::greater<Pair>> pq;

        for (int val : arr) {
            int cnt = 0;
            for (int i = 0; i < 32; ++i) {
                int bit = ((1 << i) & val) == 0 ? 0 : 1;
                if (bit == 1) {
                    cnt++;
                }
            }
            pq.push(Pair(val, cnt));
        }

        std::vector<int> ans(arr.size());
        int ind = 0;
        while (!pq.empty()) {
            Pair rem = pq.top();
            pq.pop();
            ans[ind++] = rem.num;
        }

        return ans;
    }
};
